src/modalities/models/model_factory.py

```diff
@@ -132,3 +245,83 @@ def get_activation_checkpointed_model(model: FSDP, activation_checkpointing_modu
                     f"Current model type: {type(model)}"
                 )
         return model
+
+    @staticmethod
+    def get_compiled_model(model: nn.Module, block_names: list[str]) -> nn.Module:
+        """
+        Apply torch.compile to each transformer block, which makes compilation efficient due to
+        repeated structure. Alternatively one can compile the whole model (after applying DP).
+        Inspired by: https://github.com/pytorch/torchtitan/blob/6b2912a9b53464bfef744e62100716271b2b248f/torchtitan/parallelisms/parallelize_llama.py#L275
+        """
+
+        def get_parent_module_and_child_name(child_module: nn.Module, model: nn.Module) -> tuple[nn.Module, str]:
+            for _, parent_candidate in model.named_modules():
+                for child_name, child_candidate in parent_candidate.named_children():
+                    if child_candidate is child_module:
+                        return parent_candidate, child_name
+            raise ModelStateError("No valid parent candidate")
+
+        block_types = tuple([get_module_class_from_name(model, b) for b in block_names])
+
+        for _, module in model.named_modules():
+            if isinstance(module, block_types):
+                compiled_module = torch.compile(module, fullgraph=True)
+                parent_module, child_name = get_parent_module_and_child_name(child_module=module, model=model)
+                parent_module.register_module(name=child_name, module=compiled_module)
```

Comment:
Why do we need to register the module?

Reply:
To maximize speedup this is recommended. 
https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html


> In order to maximize speedup, graph breaks should be limited. We can force TorchDynamo to raise an error upon the first graph break encountered by using 
```
fullgraph=True:

opt_bar = torch.compile(bar, fullgraph=True)
try:
    opt_bar(torch.randn(10), torch.randn(10))
except:
    tb.print_exc()
```